allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,824 @@
+                                                  top_span_mask: torch.FloatTensor) -> torch.FloatTensor:
+        """
+        Computes the negative marginal log likelihood with respect to the gold cluster labels.
+        This computes the log of the sum of the probabilities of all antecedent predictions that
+        would be consistent with the data, in the sense that we are minimising, for a given span,
+        the negative marginal log likelihood of all antecedents which are in the same gold cluster
+        as the span we are currently considering. Each span i predicts a single antecedent j, but
+        there might be several prior mentions k in the same coreference cluster that would be valid
+        antecedents. Our loss is the sum of the probability assigned to all valid antecedents.
+        This is a valid objective for clustering as we don't mind which antecedent is predicted,
+        so long as they are in the same coreference cluster.
+
+        The computation is performed in log-space for numerical stability.
+
+        Parameters
+        ----------
+        coreference_scores: ``torch.FloatTensor``, required.
+            The pairwise scores between every span and its possible antecedents.
+            Has shape (batch_size, num_spans_to_keep, max_antecedents + 1).
+        augmented_labels: ``torch.IntTensor``, required.
+            A binary indicator for whether it is consistent with the data for a span to choose
+            an antecedent. Has shape (batch_size, num_spans_to_keep, max_antecedents + 1).
+        top_span_mask : ``torch.FloatTensor``, required.
+            A binary mask of shape (batch_size, num_spans_to_keep, 1), representing which
+            spans are valid, as we may have considered different numbers of spans per batch.
+            It is infrequent that this is not all ones, because we prune to find the top spans,
+            and generally there are at least ``num_spans_to_keep`` spans for each element in the
+            batch and additionally, the mention scorer should learn to score valid mentions
+            higher than padded ones.
+
+        Returns
+        -------
+        negative_marginal_log_likelihood: ``torch.FloatTensor``
+            The scalar negative marginal log likelihood of the gold cluster labels.
+        """
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents + 1)
+        # Adding the log of the mask means that we add -inf to all
+        # antecedents not in the coreference cluster for a given span,
+        # meaning when we compute the marginal log likelihood, these are
+        # correctly masked to have no contribution.
+        masked_coreference_scores = coreference_scores + augmented_labels.log()
+
+        # Shape: (batch_size, num_spans_to_keep)
+        marginalized_gold_scores = util.logsumexp(masked_coreference_scores, 2)
+        log_norm = util.logsumexp(coreference_scores, 2)
```

Comment:
Still pretty sure you can just do:

```python
coreference_log_probs = util.last_dim_softmax(coreference_scores, top_span_mask)
# Adding `augmented_labels.log()` masks out all coreference decisions that are inconsistent with the
# gold coreference clusters, letting us sum over all possible antecedents that are in the correct cluster.
correct_antecedent_log_probs = coreference_log_probs + augmented_labels.log()
loss = -util.logsumexp(correct_antecedent_log_probs)
```
And this is now simple enough to go in `forward` directly, even.

Reply:
Oh, the softmax takes it out of logspace.  But then you just multiple by the mask, instead of adding the log mask, and you do a sum instead of `logsumexp`.  That's probably even faster still, actually.